wgpu-core/src/lock/observing.rs

```diff
@@ -0,0 +1,330 @@
+//! acquires one lock while it is already holding another, we note
+//! that nesting pair. This tells us what the [`LockRank::followers`]
+//! set for each lock would need to include to accommodate
+//! `wgpu-core`'s observed behavior.
+//!
+//! When `wgpu-core`'s `observe_locks` feature is enabled, if the
+//! `WGPU_CORE_LOCK_OBSERVE_DIR` environment variable is set to the
+//! path of an existing directory, then every thread that acquires a
+//! lock in `wgpu-core` will write its own log file to that directory.
+//! You can then run the `wgpu` workspace's `analyze-locks` binary
+//! to read those files and summarize the results. The output from
+//! `analyze-locks` has the same form as the lock ranks given in
+//! [`lock/rank.rs`].
+//!
+//! If the `WGPU_CORE_LOCK_OBSERVE_DIR` environment variable is not
+//! set, then no instrumentation takes place, and the locks behave
+//! normally.
+//!
+//! To make sure we capture all acquisitions regardless of when the
+//! program exits, each thread writes events directly to its log file
+//! as they occur. A `write` system call is generally just a copy from
+//! userspace into the kernel's buffer, so hopefully this approach
+//! will still have tolerable performance.
+//!
+//! [`lock/rank.rs`]: ../../../src/lock/rank.rs
+
+#![allow(dead_code)]
+
+use crate::FastHashSet;
+
+use super::rank::{LockRank, LockRankSet};
+use std::{cell::RefCell, panic::Location, path::PathBuf};
+
+/// A `Mutex` instrumented for lock acquisition order observation.
+///
+/// This is just a wrapper around a `parking_lot::Mutex`, along with
+/// its rank in the `wgpu_core` lock ordering.
+///
+/// For details, see [the module documentation][mod].
+///
+/// [mod]: crate::lock::observing
+pub struct Mutex<T> {
+    inner: parking_lot::Mutex<T>,
+    rank: LockRank,
+}
+
+/// A guard produced by locking [`Mutex`].
+///
+/// This is just a wrapper around a `parking_lot::MutexGuard`, along
+/// with the state needed to track lock acquisition.
+///
+/// For details, see [the module documentation][mod].
+///
+/// [mod]: crate::lock::observing
+pub struct MutexGuard<'a, T> {
+    inner: parking_lot::MutexGuard<'a, T>,
+
+    /// The youngest lock that was already held when we acquired this
+    /// one, if any.
+    saved: Option<HeldLock>,
+}
+
+thread_local! {
+    static LOCK_STATE: RefCell<ThreadState> = const { RefCell::new(ThreadState::Initial) };
+}
+
+/// Thread-local state for lock observation.
+enum ThreadState {
+    /// This thread hasn't yet checked the environment variable.
+    Initial,
+
+    /// This thread checked the environment variable, and it was
+    /// unset, so this thread is not observing lock acquisitions.
+    Disabled,
+
+    /// Lock observation is enabled for this thread.
+    Enabled(ObservationState),
+}
+
+struct ObservationState {
+    /// Log lock observations to this file.
+    log_file: std::fs::File,
+
+    /// Locations we've seen so far.
```

Comment:
nitpick(non-blocking): just for funsies:

```suggestion
    /// [`Location`]s we've seen so far.
```